### PR TITLE
task4

### PR DESCRIPTION
**Pull request by Paul044**
task4

### CODE REVIEW

**Comment by Paul044** [comment]
As user table has soft delete support I added a transaction to clean userGroup entities, for case with deletion of groups entities ( which is not paranoid: true) sequelize usesdefault onDelete: 'CASCADE' for models.

Had some issue with error handling, left a comment in the code.

File: task4/src/models/index.js [hunk 1]
```diff
@@ -1 +1,24 @@
...
+    });
+
+Object.keys(db).forEach((modelName) => {
+    if (typeof db[modelName].associate === 'function') {
+        db[modelName].associate(db);
+    }
+});
+
+sequelize.sync();
+
```
Comment (meteorra) [inline task4/src/models/index.js:23]: What is this code for? If it's needed to sync data in DBs, then usually it's done with deployment scripts, not in terms of application

File: task4/src/routes/group/index.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        } catch (err) {
+            res.status(err.status || 500).json({
+                message: err
+            });
+        }
+    })
+    .delete(async (req, res) => {
+        try {
+            const { id: groupId } = req.params;
+            const group = await groupService.getById(groupId);
```
Comment (meteorra) [inline task4/src/routes/group/index.js]: No need to search for a group & then delete, you can just delete & catch error

File: task4/src/routes/group/index.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                message: err
+            });
+        }
+    });
+
+router
+    .route('/')
+    .get(async (req, res) => {
+        try {
+            const Group = await groupService.getAllGroups();
```
Comment (meteorra) [inline task4/src/routes/group/index.js]: Why this group is capital case?

File: task4/src/routes/group/index.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        } catch (err) {
+            res.status(err.status || 500).json({
+                message: err
+            });
+        }
+    })
+    .put(async (req, res) => {
+        try {
+            const group = req.body;
+            const storedUser = await groupService.getById(group.id);
```
Comment (meteorra) [inline task4/src/routes/group/index.js:83]: No need to search for a group & then delete, you can just delete & catch error

File: task4/src/services/groupService.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
+export default class GroupService {
+    constructor(models) {
+        this.model = models.Group;
+        this.userModel = models.User;
+    }
+
+    getAllGroups() {
+        try {
+            const data = this.model.findAll({
```
Comment (meteorra) [inline task4/src/services/groupService.js:9]:
```
return this.model.findAll({
                include: [
                    {
                        model: this.userModel,
                        as: 'users',
                        required: false,
                        attributes: ['id', 'login'],
                        through: { attributes: [] }
                    }
                ]
            })

No need to wrap in try catch here,  a promise can be returned

File: task4/src/services/groupService.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                        as: 'users',
+                        required: false,
+                        attributes: ['id', 'login'],
+                        through: { attributes: [] }
+                    }
+                ]
+            });
+            return data;
+        } catch (err) {
+            throw err;
```
Comment (meteorra) [inline task4/src/services/groupService.js:22]: BTW,  what will be caught here in catch without await?

File: task4/src/models/index.js (hunk 1 above)
Comment (Paul044) [reply task4/src/models/index.js:23]: It allows to create tables, if they aren't exist. And used to add {force: true} there, when I needed to add or modify columns in my tables.

File: task4/src/routes/group/index.js (hunk 3 above)
Comment (Paul044) [reply task4/src/routes/group/index.js]: I copied user routes and replaced it with group title, something must have gone wrong, I'll fix it

File: task4/src/routes/group/index.js (hunk 2 above)
Comment (Paul044) [reply task4/src/routes/group/index.js]: As discussed with @meteorra, it's OK to receive success response for deleting  not existed item, so I'm getting rid of this code

File: task4/src/services/groupService.js (hunk 5 above)
Comment (Paul044) [reply task4/src/services/groupService.js:9]: So, am I understand correctly, that we can rely on all synchronous model's methods, and we can just return promise for all services methods ?

File: task4/src/services/groupService.js (hunk 6 above)
Comment (Paul044) [reply task4/src/services/groupService.js:22]: Is it possible that model's method can throw an error?

**Pull request merged**